File: src/element.cpp
#include "element.h"

#include <sstream>

Element::~Element() {
  children.clear();
  gtk_widget_destroy(widget);
}

Element* Element::add(std::unique_ptr<Element>&& element) {
  gtk_container_add(GTK_CONTAINER(widget), element->widget);
  element->visible();
  children.emplace_back(std::move(element));
  return this;
}

Element* Element::visible(bool value) {
  gtk_widget_set_visible(widget, value);
  return this;
}

Element* Element::addClass(const std::string& classNames) {
  GtkStyleContext* style = gtk_widget_get_style_context(widget);
  std::istringstream iss(classNames);
  std::string name;
  while (std::getline(iss, name, ' '))
    gtk_style_context_add_class(style, name.c_str());
  return this;
}

Element* Element::removeClass(const std::string& className) {
  gtk_style_context_remove_class(gtk_widget_get_style_context(widget),
                                 className.c_str());
  return this;
}

Element* Element::size(int16_t width, int16_t height) {
  gtk_widget_set_size_request(widget, width, height);
  return this;
}

Element* Element::tooltip(const std::string& text) {
  gtk_widget_set_tooltip_markup(widget, text.c_str());
  return this;
}

Element* Element::focus() {
  gtk_widget_grab_focus(widget);
  return this;
}

Element* Element::addState(GtkStateFlags flag) {
  GtkStateFlags flags = gtk_widget_get_state_flags(widget);
  if (!(flags & flag)) gtk_widget_set_state_flags(widget, flag, false);
  return this;
}

Element* Element::removeState(GtkStateFlags flag) {
  GtkStateFlags flags = gtk_widget_get_state_flags(widget);
  if (flags & flag) gtk_widget_unset_state_flags(widget, flag);
  return this;
}

PointerEvents* PointerEvents::onPointerDown(const PointerCallback& callback) {
  pointerDownCallback = callback;
  gtk_widget_add_events(widget, GDK_BUTTON_PRESS_MASK);
  g_signal_connect(widget, "button-press-event",
                   G_CALLBACK(+[](GtkWidget*, GdkEventButton* event,
                                  gpointer data) -> gboolean {
                     PointerEvents* _this = static_cast<PointerEvents*>(data);
                     _this->pointerDownCallback(event);
                     return GDK_EVENT_PROPAGATE;
                   }),
                   this);
  return this;
}

PointerEvents* PointerEvents::onPointerUp(const PointerCallback& callback) {
  pointerUpCallback = callback;
  auto pressed = [](GtkWidget*, GdkEventButton* event,
                    gpointer data) -> gboolean {
    PointerEvents* _this = static_cast<PointerEvents*>(data);
    _this->pointerUpCallback(event);
    return GDK_EVENT_PROPAGATE;
  };
  gtk_widget_add_events(widget, GDK_BUTTON_RELEASE_MASK);
  g_signal_connect(widget, "button-release-event", G_CALLBACK(+pressed), this);
  return this;
}

ScrollEvents* ScrollEvents::onScroll(
    const std::function<void(ScrollDirection)>& callback) {
  scrollCallback = callback;
  auto scroll = [](GtkWidget*, GdkEventScroll* event,
                   gpointer data) -> gboolean {
    ScrollEvents* _this = static_cast<ScrollEvents*>(data);
    if (event->direction == GDK_SCROLL_UP) {
      _this->scrollCallback(ScrollDirection::Up);
    } else if (event->direction == GDK_SCROLL_DOWN) {
      _this->scrollCallback(ScrollDirection::Down);
    } else if (event->direction == GDK_SCROLL_SMOOTH) {
      _this->scrollCallback(event->delta_y > 0 ? ScrollDirection::Down
                                               : ScrollDirection::Up);
    }
    return GDK_EVENT_PROPAGATE;
  };
  gtk_widget_add_events(widget, GDK_SCROLL_MASK);
  g_signal_connect(widget, "scroll-event", G_CALLBACK(+scroll), this);
  return this;
}

gboolean HoverEvents::onHoverChange(GtkWidget*, GdkEventCrossing* event,
                                    gpointer data) {
  bool self = event->detail == GDK_NOTIFY_NONLINEAR ||
              event->detail == GDK_NOTIFY_NONLINEAR_VIRTUAL;
  HoverEvents* _this = static_cast<HoverEvents*>(data);
  if (event->type == GDK_ENTER_NOTIFY) _this->hoverCallback(self);
  if (event->type == GDK_LEAVE_NOTIFY) _this->hoverOutCallback(self);
  return GDK_EVENT_PROPAGATE;
}

HoverEvents* HoverEvents::onHover(const HoverCallback& callback) {
  hoverCallback = callback;
  gtk_widget_add_events(widget, GDK_ENTER_NOTIFY_MASK);
  g_signal_connect(widget, "enter-notify-event", (GCallback)onHoverChange,
                   this);
  return this;
}

HoverEvents* HoverEvents::onHoverOut(const HoverCallback& callback) {
  hoverOutCallback = callback;
  gtk_widget_add_events(widget, GDK_LEAVE_NOTIFY_MASK);
  g_signal_connect(widget, "leave-notify-event", (GCallback)onHoverChange,
                   this);
  return this;
}

KeyboardEvents* KeyboardEvents::onKeyDown(const Callback& callback) {
  keyDownCallback = callback;
  g_signal_connect(widget, "key-press-event",
                   G_CALLBACK(+[](GtkWidget*, GdkEventKey* event,
                                  gpointer data) -> gboolean {
                     KeyboardEvents* _this = static_cast<KeyboardEvents*>(data);
                     _this->keyDownCallback(event);
                     return GDK_EVENT_PROPAGATE;
                   }),
                   this);
  return this;
}

VisibilityEvents* VisibilityEvents::onHide(
    const std::function<void()>& callback) {
  hideCallback = callback;
  auto hide = [](GtkWidget* widget, gpointer data) -> gboolean {
    VisibilityEvents* _this = static_cast<VisibilityEvents*>(data);
    _this->hideCallback();
    return GDK_EVENT_PROPAGATE;
  };
  g_signal_connect(widget, "hide", G_CALLBACK(+hide), this);
  return this;
}

Box::Box(GtkOrientation orientation) {
  widget = gtk_box_new(orientation, 0);
  spaceEvenly(false);
}

Box* Box::gap(std::uint16_t value) {
  gtk_box_set_spacing((GtkBox*)widget, value);
  return this;
}

Box* Box::spaceEvenly(bool value) {
  gtk_box_set_homogeneous((GtkBox*)widget, value);
  return this;
}

Box* Box::prependChild(std::unique_ptr<Element>&& child) {
  gtk_box_pack_start((GtkBox*)widget, child->widget, true, true, 0);
  child->visible();
  children.emplace_back(std::move(child));
  return this;
}

Label::Label(const std::string& value) {
  widget = gtk_label_new(value.c_str());
  // gtk_label_set_ellipsize((GtkLabel *)widget, PANGO_ELLIPSIZE_END);
}

Label* Label::set(const std::string& value) {
  gtk_label_set_text((GtkLabel*)widget, value.c_str());
  return this;
}

Icon::Icon() : Box(GTK_ORIENTATION_HORIZONTAL) { addClass("icon"); }

Icon* Icon::set(const std::string& name) {
  if (!label) {
    auto _label = std::make_unique<Label>();
    label = _label.get();
    add(std::move(_label));
  }
  label->set(name);
  return this;
}

Icon* Icon::setImage(const std::string& path) {
  GtkStyleContext* context = gtk_widget_get_style_context(widget);
  if (!gtk_style_context_has_class(context, "image")) addClass("image");
  if (!style) style = std::make_unique<Style>(widget);
  style->css("* { background-image: url(\"" + path + "\"); }");
  return this;
}

Button::Button(Type type, Variant variant, Size size) {
  widget = gtk_button_new();
  if (variant == Filled)
    addClass("filled");
  else if (variant != Tonal) {
    // GTK adds "flat" class.
    gtk_button_set_relief((GtkButton*)widget, GTK_RELIEF_NONE);
  }
  if (size == Small) addClass("small");

  // Don't stretch children.
  gtk_widget_set_valign(widget, GTK_ALIGN_CENTER);

  auto _container = std::make_unique<Box>();
  container = _container.get();

  if (type == Type::Icon) {
    addClass("icon circle");
    gtk_widget_set_halign(container->widget, GTK_ALIGN_CENTER);
  }

  if (type == Type::IconText) {
    auto _startIcon = std::make_unique<Icon>();
    startIcon = _startIcon.get();
    _startIcon->addClass("start-icon");
    container->add(std::move(_startIcon));
  }

  auto _content = std::make_unique<Box>();
  content = _content.get();
  container->add(std::move(_content));

  if (type == Type::IconText) {
    auto _endIcon = std::make_unique<Icon>();
    endIcon = _endIcon.get();
    _endIcon->addClass("end-icon");
    container->add(std::move(_endIcon));
  }

  add(std::move(_container));
}

Button* Button::setContent(std::unique_ptr<Element>&& element) {
  content->children.clear();
  content->add(std::move(element));
  return this;
}

Button* Button::setContent(const std::string& value) {
  auto label = std::make_unique<Label>(value);
  gtk_widget_set_halign(label->widget, GTK_ALIGN_START);
  setContent(std::move(label));
  return this;
}

Button* Button::onClick(const std::function<void()>& callback) {
  clickCallback = callback;
  auto clicked = [](GtkButton* button, gpointer data) -> gboolean {
    Button* _this = static_cast<Button*>(data);
    _this->clickCallback();
    return GDK_EVENT_STOP;
  };
  g_signal_connect(widget, "clicked", G_CALLBACK(+clicked), this);
  return this;
}

bool Button::disabled() { return !gtk_widget_get_sensitive(widget); }

Button* Button::disabled(bool value) {
  gtk_widget_set_sensitive(widget, !value);
  return this;
}

ScrolledWindow::ScrolledWindow() {
  widget = gtk_scrolled_window_new(nullptr, nullptr);
  gtk_scrolled_window_set_policy((GtkScrolledWindow*)widget,
                                 GTK_POLICY_AUTOMATIC, GTK_POLICY_AUTOMATIC);
  gtk_widget_set_vexpand(widget, true);
}

Input::Input() {
  widget = gtk_entry_new();
  gtk_widget_set_hexpand(widget, true);
}

std::string Input::value() { return gtk_entry_get_text((GtkEntry*)widget); }

Input* Input::value(const std::string& value) {
  gtk_entry_set_text((GtkEntry*)widget, value.c_str());
  return this;
}

Input* Input::placeholder(const std::string& value) {
  gtk_entry_set_placeholder_text((GtkEntry*)widget, value.c_str());
  return this;
}

Input* Input::onChange(const std::function<void()>& callback) {
  changeCallback = callback;
  g_signal_connect(widget, "changed",
                   G_CALLBACK(+[](GtkWidget*, gpointer data) {
                     auto _this = static_cast<Input*>(data);
                     _this->changeCallback();
                   }),
                   this);
  return this;
}

Input* Input::onSubmit(const std::function<void()>& callback) {
  submitCallback = callback;
  g_signal_connect(widget, "activate",
                   G_CALLBACK(+[](GtkWidget*, gpointer data) {
                     auto _this = static_cast<Input*>(data);
                     _this->submitCallback();
                   }),
                   this);
  return this;
}

Slider::Slider() {
  widget = gtk_scale_new_with_range(GTK_ORIENTATION_HORIZONTAL, 0, 100, 1);
  gtk_scale_set_draw_value((GtkScale*)widget, false);
  gtk_widget_set_hexpand(widget, true);
}

uint8_t Slider::value() { return gtk_range_get_value(GTK_RANGE(widget)); }

Slider* Slider::value(uint8_t value) {
  gtk_range_set_value(GTK_RANGE(widget), value);
  return this;
}

Slider* Slider::onChange(const std::function<void()>& callback) {
  changeCallback = callback;
  auto changed = [](GtkRange* range, gpointer data) {
    Slider* _this = static_cast<Slider*>(data);
    _this->changeCallback();
    return GDK_EVENT_PROPAGATE;
  };
  g_signal_connect(widget, "value-changed", G_CALLBACK(+changed), this);
  return this;
}

FlowBoxChild::FlowBoxChild() { widget = gtk_flow_box_child_new(); }

FlowBox::FlowBox() {
  widget = gtk_flow_box_new();
  spaceEvenly(true);
}

FlowBox* FlowBox::gap(std::uint16_t value) {
  gtk_flow_box_set_column_spacing((GtkFlowBox*)widget, value);
  gtk_flow_box_set_row_spacing((GtkFlowBox*)widget, value);
  return this;
}

FlowBox* FlowBox::spaceEvenly(bool value) {
  gtk_flow_box_set_homogeneous((GtkFlowBox*)widget, value);
  return this;
}

FlowBox* FlowBox::columns(std::uint8_t value) {
  gtk_flow_box_set_min_children_per_line((GtkFlowBox*)widget, value);
  gtk_flow_box_set_max_children_per_line((GtkFlowBox*)widget, value);
  return this;
}

FlowBox* FlowBox::onChildClick(const ChildCallback& callback) {
  childClickCallback = callback;
  g_signal_connect(
      widget, "child-activated",
      G_CALLBACK(+[](GtkFlowBox*, GtkFlowBoxChild* childWidget, gpointer data) {
        auto _this = static_cast<FlowBox*>(data);
        _this->childClickCallback(childWidget);
      }),
      this);
  return this;
}

FlowBoxChild* FlowBox::add(std::unique_ptr<Element>&& element) {
  auto child = std::make_unique<FlowBoxChild>();
  auto ptr = child.get();
  child->add(std::move(element));
  Element::add(std::move(child));
  return ptr;
}

EventBox::EventBox() { widget = gtk_event_box_new(); }

Window::Window(GtkWindowType type, GtkLayerShellKeyboardMode keyboardMode) {
  widget = gtk_window_new(type);
  gtk_layer_init_for_window((GtkWindow*)widget);
  gtk_layer_set_layer((GtkWindow*)widget, GTK_LAYER_SHELL_LAYER_TOP);
  gtk_layer_set_keyboard_mode((GtkWindow*)widget, keyboardMode);
}

GtkLayerShellEdge gtkLayerEnumFromAlign(Align value) {
  if (value == Align::Top) return GTK_LAYER_SHELL_EDGE_TOP;
  if (value == Align::Bottom) return GTK_LAYER_SHELL_EDGE_BOTTOM;
  if (value == Align::End) return GTK_LAYER_SHELL_EDGE_RIGHT;
  return GTK_LAYER_SHELL_EDGE_LEFT;
}

Window* Window::align(Align horizontal, Align vertical) {
  gtk_layer_set_anchor((GtkWindow*)widget, gtkLayerEnumFromAlign(horizontal),
                       true);
  gtk_layer_set_anchor((GtkWindow*)widget, gtkLayerEnumFromAlign(vertical),
                       true);
  return this;
}

std::tuple<Align, Align> Window::align() {
  Align horizontal;
  Align vertical;
  if (gtk_layer_get_anchor((GtkWindow*)widget, GTK_LAYER_SHELL_EDGE_LEFT))
    horizontal = Align::Start;
  if (gtk_layer_get_anchor((GtkWindow*)widget, GTK_LAYER_SHELL_EDGE_RIGHT))
    horizontal = Align::End;
  if (gtk_layer_get_anchor((GtkWindow*)widget, GTK_LAYER_SHELL_EDGE_TOP))
    vertical = Align::Top;
  if (gtk_layer_get_anchor((GtkWindow*)widget, GTK_LAYER_SHELL_EDGE_BOTTOM))
    vertical = Align::Bottom;
  return std::make_tuple(horizontal, vertical);
}

MenuItem::MenuItem(const std::string& label, const std::string& icon) {
  widget = gtk_menu_item_new();

  auto box = std::make_unique<Box>();
  if (!icon.empty()) {
    auto _icon = std::make_unique<Icon>();
    _icon->set(icon);
    _icon->addClass("start-icon");
    box->add(std::move(_icon));
  }
  auto _label = std::make_unique<Label>();
  _label->set(label);
  box->add(std::move(_label));

  add(std::move(box));
}

MenuItem* MenuItem::onClick(const std::function<void()>& callback) {
  clickCallback = callback;
  g_signal_connect_swapped(widget, "activate", G_CALLBACK(+[](MenuItem* _this) {
                             if (_this->clickCallback) _this->clickCallback();
                           }),
                           this);
  return this;
}

MenuSeparator::MenuSeparator() { widget = gtk_separator_menu_item_new(); }

Menu::Menu() { widget = gtk_menu_new(); }

Menu* Menu::add(std::unique_ptr<Element>&& child) {
  gtk_menu_shell_append((GtkMenuShell*)widget, child->widget);
  child->visible();
  children.emplace_back(std::move(child));
  return this;
}

Menu* Menu::visible(bool value) {
  if (value) gtk_menu_popup_at_pointer((GtkMenu*)widget, nullptr);
  return this;
}

gboolean Transition::update(gpointer data) {
  Transition* _this = static_cast<Transition*>(data);
  _this->currentSteps--;
  if (_this->currentSteps > 0) {
    _this->current.width += _this->stepWidth;
    _this->current.height += _this->stepHeight;
    _this->element->size(_this->current.width, _this->current.height);
    return G_SOURCE_CONTINUE;
  } else {
    for (const auto& child : _this->element->children) child->visible();
    _this->element->size(-1, -1);  // revert dynamic sizing
    _this->finishCallback();
    return G_SOURCE_REMOVE;
  }
}

Transition* Transition::to(Frame to, const std::function<void()>& onFinish) {
  if (currentSteps > 0) {
    g_source_remove(timeout);
  } else {
    current.width = gtk_widget_get_allocated_width(element->widget);
    current.height = gtk_widget_get_allocated_height(element->widget);
    // children prevents parent size change. so hide while transitioning.
    for (const auto& child : element->children) child->visible(false);
  }
  finishCallback = onFinish;
  currentSteps = _duration / timeoutMs;
  stepWidth = (to.width - current.width) / currentSteps;
  stepHeight = (to.height - current.height) / currentSteps;
  timeout = g_timeout_add(timeoutMs, update, this);
  return this;
}

Transition::Transition(Element* element) : element(element) {}

Transition* Transition::duration(int16_t value) {
  this->_duration = value;
  return this;
}

File: zig-app/src/element.zig
const std = @import("std");
const c = @import("gtk.zig");

// ── Element (base) ─────────────────────────────────────────────────────────

pub const Element = struct {
    widget: *c.GtkWidget,
    children: std.ArrayList(*Element),
    alloc: std.mem.Allocator,

    pub fn initWith(alloc: std.mem.Allocator, widget: *c.GtkWidget) Element {
        return .{ .widget = widget, .children = std.ArrayList(*Element).init(alloc), .alloc = alloc };
    }

    pub fn deinit(self: *Element) void {
        for (self.children.items) |child| {
            child.deinit();
            self.alloc.destroy(child);
        }
        self.children.deinit();
        c.gtk_widget_destroy(self.widget);
    }

    pub fn add(self: *Element, child: *Element) *Element {
        c.gtk_container_add(@ptrCast(self.widget), child.widget);
        _ = c.gtk_widget_show(child.widget);
        self.children.append(child) catch {};
        return self;
    }

    pub fn visible(self: *Element, value: bool) *Element {
        c.gtk_widget_set_visible(self.widget, if (value) 1 else 0);
        return self;
    }

    pub fn show(self: *Element) *Element {
        c.gtk_widget_show(self.widget);
        return self;
    }

    pub fn addClass(self: *Element, classNames: []const u8) *Element {
        const ctx = c.gtk_widget_get_style_context(self.widget);
        var it = std.mem.tokenizeScalar(u8, classNames, ' ');
        var buf: [128]u8 = undefined;
        while (it.next()) |name| {
            const z = std.fmt.bufPrintZ(&buf, "{s}", .{name}) catch continue;
            c.gtk_style_context_add_class(ctx, z);
        }
        return self;
    }

    pub fn removeClass(self: *Element, className: []const u8) *Element {
        const ctx = c.gtk_widget_get_style_context(self.widget);
        var buf: [128]u8 = undefined;
        const z = std.fmt.bufPrintZ(&buf, "{s}", .{className}) catch return self;
        c.gtk_style_context_remove_class(ctx, z);
        return self;
    }

    pub fn size(self: *Element, w: i16, h: i16) *Element {
        c.gtk_widget_set_size_request(self.widget, w, h);
        return self;
    }

    pub fn focus(self: *Element) *Element {
        c.gtk_widget_grab_focus(self.widget);
        return self;
    }

    pub fn addState(self: *Element, flag: c.GtkStateFlags) *Element {
        const flags = c.gtk_widget_get_state_flags(self.widget);
        if (flags & @intFromEnum(flag) == 0)
            c.gtk_widget_set_state_flags(self.widget, flag, 0);
        return self;
    }

    pub fn removeState(self: *Element, flag: c.GtkStateFlags) *Element {
        const flags = c.gtk_widget_get_state_flags(self.widget);
        if (flags & @intFromEnum(flag) != 0)
            c.gtk_widget_unset_state_flags(self.widget, flag);
        return self;
    }
};

// ── Box ────────────────────────────────────────────────────────────────────

pub const Box = struct {
    base: Element,

    pub fn init(alloc: std.mem.Allocator, orientation: c.GtkOrientation) Box {
        const w = c.gtk_box_new(orientation, 0);
        c.gtk_box_set_homogeneous(@ptrCast(w), 0);
        return .{ .base = Element.initWith(alloc, w.?) };
    }

    pub fn gap(self: *Box, value: u16) *Box {
        c.gtk_box_set_spacing(@ptrCast(self.base.widget), @intCast(value));
        return self;
    }

    pub fn spaceEvenly(self: *Box, value: bool) *Box {
        c.gtk_box_set_homogeneous(@ptrCast(self.base.widget), if (value) 1 else 0);
        return self;
    }

    pub fn prependChild(self: *Box, child: *Element) *Box {
        c.gtk_box_pack_start(@ptrCast(self.base.widget), child.widget, 1, 1, 0);
        _ = c.gtk_widget_show(child.widget);
        self.base.children.append(child) catch {};
        return self;
    }
};

// ── Label ──────────────────────────────────────────────────────────────────

pub const Label = struct {
    base: Element,

    pub fn init(alloc: std.mem.Allocator, text: []const u8) Label {
        var buf: [512]u8 = undefined;
        const z = std.fmt.bufPrintZ(&buf, "{s}", .{text}) catch buf[0..1 :0];
        return .{ .base = Element.initWith(alloc, c.gtk_label_new(z).?) };
    }

    pub fn set(self: *Label, text: []const u8) *Label {
        var buf: [512]u8 = undefined;
        const z = std.fmt.bufPrintZ(&buf, "{s}", .{text}) catch return self;
        c.gtk_label_set_text(@ptrCast(self.base.widget), z);
        return self;
    }
};

// ── Icon ───────────────────────────────────────────────────────────────────

pub const Icon = struct {
    box: Box,
    label: ?*Label = null,

    pub fn init(alloc: std.mem.Allocator) Icon {
        var icon = Icon{ .box = Box.init(alloc, c.GTK_ORIENTATION_HORIZONTAL) };
        _ = icon.box.base.addClass("icon");
        return icon;
    }

    pub fn set(self: *Icon, name: []const u8) *Icon {
        if (self.label == null) {
            const lbl = self.box.base.alloc.create(Label) catch return self;
            lbl.* = Label.init(self.box.base.alloc, "");
            _ = self.box.base.add(&lbl.base);
            self.label = lbl;
        }
        _ = self.label.?.set(name);
        return self;
    }

    pub fn setImage(self: *Icon, path: []const u8) *Icon {
        const ctx = c.gtk_widget_get_style_context(self.box.base.widget);
        if (c.gtk_style_context_has_class(ctx, "image") == 0)
            _ = self.box.base.addClass("image");
        var cssBuf: [1024]u8 = undefined;
        const css = std.fmt.bufPrintZ(&cssBuf, "* {{ background-image: url(\"{s}\"); }}", .{path}) catch return self;
        applyCssToWidget(self.box.base.widget, css);
        return self;
    }
};

fn applyCssToWidget(widget: *c.GtkWidget, css: [:0]const u8) void {
    const provider = c.gtk_css_provider_new();
    defer c.g_object_unref(provider);
    _ = c.gtk_css_provider_load_from_data(provider, css, -1, null);
    const ctx = c.gtk_widget_get_style_context(widget);
    c.gtk_style_context_add_provider(ctx, @ptrCast(provider), c.GTK_STYLE_PROVIDER_PRIORITY_APPLICATION);
}

// ── Input ──────────────────────────────────────────────────────────────────

pub const Input = struct {
    base: Element,
    changeCallback: ?*const fn (data: ?*anyopaque) void = null,
    submitCallback: ?*const fn (data: ?*anyopaque) void = null,
    userData: ?*anyopaque = null,

    pub fn init(alloc: std.mem.Allocator) Input {
        const w = c.gtk_entry_new().?;
        c.gtk_widget_set_hexpand(w, 1);
        return .{ .base = Element.initWith(alloc, w) };
    }

    pub fn getValue(self: *Input) []const u8 {
        return std.mem.span(c.gtk_entry_get_text(@ptrCast(self.base.widget)));
    }

    pub fn setValue(self: *Input, val: []const u8) *Input {
        var buf: [512]u8 = undefined;
        const z = std.fmt.bufPrintZ(&buf, "{s}", .{val}) catch return self;
        c.gtk_entry_set_text(@ptrCast(self.base.widget), z);
        return self;
    }

    pub fn onChangeFn(self: *Input, cb: *const fn (data: ?*anyopaque) void, data: ?*anyopaque) *Input {
        self.changeCallback = cb;
        self.userData = data;
        _ = c.g_signal_connect_data(self.base.widget, "changed", @ptrCast(&inputChangedCb), data, null, 0);
        return self;
    }

    pub fn onSubmitFn(self: *Input, cb: *const fn (data: ?*anyopaque) void, data: ?*anyopaque) *Input {
        self.submitCallback = cb;
        _ = c.g_signal_connect_data(self.base.widget, "activate", @ptrCast(&inputActivateCb), data, null, 0);
        return self;
    }
};

fn inputChangedCb(_: *c.GtkWidget, data: ?*anyopaque) callconv(.C) void {
    _ = data;
}
fn inputActivateCb(_: *c.GtkWidget, data: ?*anyopaque) callconv(.C) void {
    _ = data;
}

// ── FlowBoxChild / FlowBox ─────────────────────────────────────────────────

pub const FlowBoxChild = struct {
    base: Element,

    pub fn init(alloc: std.mem.Allocator) FlowBoxChild {
        return .{ .base = Element.initWith(alloc, c.gtk_flow_box_child_new().?) };
    }
};

pub const FlowBox = struct {
    base: Element,
    childClickCb: ?*const fn (child: *c.GtkFlowBoxChild, data: ?*anyopaque) void = null,
    cbData: ?*anyopaque = null,

    pub fn init(alloc: std.mem.Allocator) FlowBox {
        const w = c.gtk_flow_box_new().?;
        c.gtk_flow_box_set_homogeneous(@ptrCast(w), 1);
        return .{ .base = Element.initWith(alloc, w) };
    }

    pub fn columns(self: *FlowBox, n: u8) *FlowBox {
        c.gtk_flow_box_set_min_children_per_line(@ptrCast(self.base.widget), n);
        c.gtk_flow_box_set_max_children_per_line(@ptrCast(self.base.widget), n);
        return self;
    }

    pub fn onChildClickFn(self: *FlowBox, cb: *const fn (*c.GtkFlowBoxChild, ?*anyopaque) void, data: ?*anyopaque) *FlowBox {
        self.childClickCb = cb;
        self.cbData = data;
        _ = c.g_signal_connect_data(self.base.widget, "child-activated", @ptrCast(&flowChildActivated), data, null, 0);
        return self;
    }

    pub fn addChild(self: *FlowBox, child: *Element) *FlowBoxChild {
        const alloc = self.base.alloc;
        const fbc = alloc.create(FlowBoxChild) catch unreachable;
        fbc.* = FlowBoxChild.init(alloc);
        _ = fbc.base.add(child);
        c.gtk_container_add(@ptrCast(self.base.widget), fbc.base.widget);
        _ = c.gtk_widget_show(fbc.base.widget);
        self.base.children.append(&fbc.base) catch {};
        return fbc;
    }
};

fn flowChildActivated(_: *c.GtkFlowBox, child: *c.GtkFlowBoxChild, data: ?*anyopaque) callconv(.C) void {
    _ = data;
    _ = child;
}

// ── ScrolledWindow ─────────────────────────────────────────────────────────

pub const ScrolledWindow = struct {
    base: Element,

    pub fn init(alloc: std.mem.Allocator) ScrolledWindow {
        const w = c.gtk_scrolled_window_new(null, null).?;
        c.gtk_scrolled_window_set_policy(@ptrCast(w), c.GTK_POLICY_AUTOMATIC, c.GTK_POLICY_AUTOMATIC);
        c.gtk_widget_set_vexpand(w, 1);
        return .{ .base = Element.initWith(alloc, w) };
    }
};

// ── EventBox ───────────────────────────────────────────────────────────────

pub const EventBox = struct {
    base: Element,

    pub fn init(alloc: std.mem.Allocator) EventBox {
        return .{ .base = Element.initWith(alloc, c.gtk_event_box_new().?) };
    }

    pub fn onPointerDown(self: *EventBox, cb: c.GCallback, data: ?*anyopaque) *EventBox {
        c.gtk_widget_add_events(self.base.widget, c.GDK_BUTTON_PRESS_MASK);
        _ = c.g_signal_connect_data(self.base.widget, "button-press-event", cb, data, null, 0);
        return self;
    }

    pub fn onHover(self: *EventBox, cb: c.GCallback, data: ?*anyopaque) *EventBox {
        c.gtk_widget_add_events(self.base.widget, c.GDK_ENTER_NOTIFY_MASK);
        _ = c.g_signal_connect_data(self.base.widget, "enter-notify-event", cb, data, null, 0);
        return self;
    }

    pub fn onHoverOut(self: *EventBox, cb: c.GCallback, data: ?*anyopaque) *EventBox {
        c.gtk_widget_add_events(self.base.widget, c.GDK_LEAVE_NOTIFY_MASK);
        _ = c.g_signal_connect_data(self.base.widget, "leave-notify-event", cb, data, null, 0);
        return self;
    }
};

// ── Window ─────────────────────────────────────────────────────────────────

pub const KeyboardMode = enum(c_int) {
    None = 0,
    Exclusive = 1,
    OnDemand = 2,
};

pub const Window = struct {
    base: EventBox,

    pub fn init(alloc: std.mem.Allocator, keyboardMode: KeyboardMode) Window {
        const w = c.gtk_window_new(c.GTK_WINDOW_TOPLEVEL).?;
        c.gtk_layer_init_for_window(@ptrCast(w));
        c.gtk_layer_set_layer(@ptrCast(w), c.GTK_LAYER_SHELL_LAYER_TOP);
        c.gtk_layer_set_keyboard_mode(@ptrCast(w), @intFromEnum(keyboardMode));
        return .{ .base = .{ .base = Element.initWith(alloc, w) } };
    }

    pub fn setNamespace(self: *Window, name: [:0]const u8) void {
        c.gtk_layer_set_namespace(@ptrCast(self.base.base.widget), name);
    }

    pub fn onKeyDown(self: *Window, cb: c.GCallback, data: ?*anyopaque) void {
        _ = c.g_signal_connect_data(self.base.base.widget, "key-press-event", cb, data, null, 0);
    }
};

// ── MenuItem / Menu ────────────────────────────────────────────────────────

pub const MenuItem = struct {
    base: Element,

    pub fn init(alloc: std.mem.Allocator, label: []const u8, icon: []const u8) MenuItem {
        const w = c.gtk_menu_item_new().?;
        var item = MenuItem{ .base = Element.initWith(alloc, w) };

        const box = alloc.create(Box) catch unreachable;
        box.* = Box.init(alloc, c.GTK_ORIENTATION_HORIZONTAL);

        if (icon.len > 0) {
            const ico = alloc.create(Icon) catch unreachable;
            ico.* = Icon.init(alloc);
            _ = ico.set(icon);
            _ = ico.box.base.addClass("start-icon");
            _ = box.base.add(&ico.box.base);
        }

        const lbl = alloc.create(Label) catch unreachable;
        lbl.* = Label.init(alloc, label);
        _ = box.base.add(&lbl.base);
        _ = item.base.add(&box.base);
        return item;
    }

    pub fn onClick(self: *MenuItem, cb: c.GCallback, data: ?*anyopaque) *MenuItem {
        _ = c.g_signal_connect_data(self.base.widget, "activate", cb, data, null, 0);
        return self;
    }
};

pub const MenuSeparator = struct {
    base: Element,
    pub fn init(alloc: std.mem.Allocator) MenuSeparator {
        return .{ .base = Element.initWith(alloc, c.gtk_separator_menu_item_new().?) };
    }
};

pub const Menu = struct {
    base: Element,

    pub fn init(alloc: std.mem.Allocator) Menu {
        return .{ .base = Element.initWith(alloc, c.gtk_menu_new().?) };
    }

    pub fn addItem(self: *Menu, child: *Element) void {
        c.gtk_menu_shell_append(@ptrCast(self.base.widget), child.widget);
        _ = c.gtk_widget_show(child.widget);
        self.base.children.append(child) catch {};
    }

    pub fn clearItems(self: *Menu) void {
        for (self.base.children.items) |child| {
            c.gtk_container_remove(@ptrCast(self.base.widget), child.widget);
        }
        self.base.children.clearRetainingCapacity();
    }

    pub fn popup(self: *Menu) void {
        c.gtk_menu_popup_at_pointer(@ptrCast(self.base.widget), null);
    }

    pub fn onHide(self: *Menu, cb: c.GCallback, data: ?*anyopaque) void {
        _ = c.g_signal_connect_data(self.base.widget, "hide", cb, data, null, 0);
    }
};
